python/pylibraft/pylibraft/neighbors/cagra/cagra.pyx

```diff
@@ -0,0 +1,852 @@
+
+    @property
+    def algo(self):
+        return self.params.algo
+
+    @property
+    def team_size(self):
+        return self.params.team_size
+
+    @property
+    def search_width(self):
+        return self.params.search_width
+
+    @property
+    def min_iterations(self):
+        return self.params.min_iterations
+
+    @property
+    def thread_block_size(self):
+        return self.params.thread_block_size
+
+    @property
+    def hashmap_mode(self):
+        return self.params.hashmap_mode
+
+    @property
+    def hashmap_min_bitlen(self):
+        return self.params.hashmap_min_bitlen
+
+    @property
+    def hashmap_max_fill_rate(self):
+        return self.params.hashmap_max_fill_rate
+
+    @property
+    def num_random_samplings(self):
+        return self.params.num_random_samplings
+
+    @property
+    def rand_xor_mask(self):
+        return self.params.rand_xor_mask
+
+
+@auto_sync_handle
+@auto_convert_output
+def search(SearchParams search_params,
+           Index index,
+           queries,
+           k,
+           neighbors=None,
+           distances=None,
+           handle=None):
+    """
+    Find the k nearest neighbors for each query.
+
+    Parameters
+    ----------
+    search_params : SearchParams
+    index : Index
+        Trained CAGRA index.
+    queries : CUDA array interface compliant matrix shape (n_samples, dim)
+        Supported dtype [float, int8, uint8]
+    k : int
+        The number of neighbors.
+    neighbors : Optional CUDA array interface compliant matrix shape
+                (n_queries, k), dtype int64_t. If supplied, neighbor
+                indices will be written here in-place. (default None)
+    distances : Optional CUDA array interface compliant matrix shape
+                (n_queries, k) If supplied, the distances to the
+                neighbors will be written here in-place. (default None)
+    {handle_docstring}
+
+    Examples
+    --------
+    >>> import cupy as cp
+
+    >>> from pylibraft.common import DeviceResources
+    >>> from pylibraft.neighbors import cagra
+
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> n_queries = 1000
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+
+    >>> # Build index
+    >>> handle = DeviceResources()
+    >>> index = cagra.build(cagra.IndexParams(), dataset, handle=handle)
+
+    >>> # Search using the built index
+    >>> queries = cp.random.random_sample((n_queries, n_features),
+    ...                                   dtype=cp.float32)
+    >>> k = 10
+    >>> search_params = cagra.SearchParams(
+    ...     max_queries=100,
+    ...     itopk_size=64,
+    ... )
+
+    >>> # Using a pooling allocator reduces overhead of temporary array
+    >>> # creation during search. This is useful if multiple searches
+    >>> # are performad with same query size.
+    >>> import rmm
+    >>> mr = rmm.mr.PoolMemoryResource(
```

Comment:
We should just remove this. I don't think this argument is supported on cagra anyways.